### DISCUSSION TITLE
compile some cells to wasm

### DISCUSSION THREAD

**Discussion by JJJHolscher** [evcxr/evcxr#367] [Ideas]
Hi, I'm using Dioxus for visualizing my rust data structures and I would want to be able to do that from within the notebook.
Dioxus needs the compilation target to be wasm however, and it appears that I cannot set a different compilation target per cell.

I take it for granted that anything that needs to be compiled to wasm, cannot share variables with anything that is to be compiled to the local binary. So, I'm asking. Is it possible to have a second kernel active (or a 1-time-use kernel per wasm cell) that does what evcxr does, but with wasm as the compilation target? Preferably it would inject the wasm (with some js glue) into the web page, so that other evcxr_display calls can reference that and use it to visualize the datastructure.

I'm willing to try to get this to work myself. But I would not know where to begin or where to look for this.

**Comment by davidlattimore** [comment]
Sounds like an interesting idea. I assume that a cell compiled to wasm would just provide functions that could be called from javascript? So it wouldn't need the same sort of wrapping that happens with regular evcxr cells?

Would wasm cells have separate dependencies from regular cells, or would they share dependencies?

I'm currently busy writing a linker, but I'm happy to mentor / provide guidance if you're keen. If you'd like to discuss via a video call, my calendar is linked from my github profile.

**Reply by JJJHolscher** [reply]
Yes, I'm unfamiliar with what normal evcxr cells need, but the wasm should just be exposed to the javascript.
Wasm cells would have separate dependencies from regular cells I imagine, since some crates don't work for the wasm target.

I myself have about 2 days to spend on getting to an mvp (and I would debug the implementation while using it for the next few days/weeks), if it would take longer then I could take a week or 2 in the summer. Therefore I _guess_ that it would be easier for wasm cells to also be isolated from each other for an initial implementation. Maybe later on, when I have more time (and feel like it would benefit my workflow) I would have wasm cells share variables. Though, if you think this aspect takes no extra resources to get, then I would immediately implement it such that wasm cells share variables.

Thank you for the offer to meet. I would take you up on it but I'm situated in Europe and am asleep during all of your available hours for the rest of March. I cannot rise earlier than 06:00+0 or stay up later than 22:00+0.
Would you be willing and able to message me the locations in the code that I would need to look at? Provided that you think you approximately understand what I intend to do, if not then I'll try to give a more complete description.

**Reply by davidlattimore** [reply]
Sharing variables between wasm cells would add a heap of complexity, so if it's not needed initially, then that's good.

I'm guessing you'd need a way to tag a cell as wasm. One option would be to have a `:wasm` command that you put at the start of the cell which changes just that cell to be wasm.

We could either have a separate command to add dependencies to wasm (e.g. `:wasm_dep`), or perhaps nicer, just say that using the `:dep` command in a wasm cell adds dependencies for wasm cells.

I'd imagine that wasm cells would go through a pretty different code path to regular cells, but can probably reuse some bits of existing code where it makes sense.

So roughly, when a `:wasm` cell is being processed, I think it'd need to do the following:

* Write the wasm code pretty much as-is to a wasm-specific directory in the evcxr temporary directory.
* Write a Cargo.toml to that directory
* Invoke cargo with appropriate flags
* Report any errors back to the user. 
* If the build succeeds, return the wasm blob evcxr_jupyter, which then sends it to the browser - I'm a little fuzzy on how you'd send wasm to the browser through jupyter, so it might make sense to experiment with that bit first.

Internally, evcxr has a struct `CodeBlock` that it uses to store some Rust code. Each line of code in the block can have a different origin. So some of the code might have come from the user and some might have been generated by evcxr. This is used when reporting errors to make sure the error gets reported in the correct location. If you do no wrapping at all, then it might not be useful to use `CodeBlock`, since the error spans reported by the compiler will correspond exactly to the user code. However if you think you might need to do some modification (e.g. adding use statements), then it might be a good idea to use it anyway. Either option is fine with me.

I've added some bookable time slots later in my evening. It's a relatively small window between when the kids are in bed and when I go to bed, but hopefully it's at least a little more European time zone friendly.

**Comment by JJJHolscher** [comment]
![image](https://github.com/evcxr/evcxr/assets/25668596/4c05ab04-2951-496f-9d17-c72f7f6e0197)
mission accomplished, I'll make a public fork sometime this week.

**Reply by JJJHolscher** [reply]
Here is the fork https://github.com/JJJHolscher/evcxr-wasm
The example notebook is converted to a web page with quarto to here https://www.mousetrap.blog/evcxr-wasm/evcxr_jupyter_tour.html#display-using-wasm which shows a working widget.
